include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1137,49 +1136,76 @@ struct __early_exit_find_or
-    // Limiting this also helps to avoid huge work-group sizes on some devices (e.g., FPGU emulation).
-    __wgroup_size = std::min(__wgroup_size, (std::size_t)2048);
-#endif
-    auto __max_cu = oneapi::dpl::__internal::__max_compute_units(__exec);
-
-    auto __n_groups = oneapi::dpl::__internal::__dpl_ceiling_div(__rng_n, __wgroup_size);
-    __n_groups = ::std::min(__n_groups, decltype(__n_groups)(__max_cu));
+    using __result_and_scratch_storage_t = __result_and_scratch_storage<_ExecutionPolicy, _AtomicType>;
+    __result_and_scratch_storage_t __result_storage(__exec, 0);
+
+    // Calculate the number of elements to be processed by each work-item.
+    const auto __iters_per_work_item = oneapi::dpl::__internal::__dpl_ceiling_div(__rng_n, __wgroup_size);
+
+    // main parallel_for
+    auto __event = __exec.queue().submit([&](sycl::handler& __cgh) {
+        oneapi::dpl::__ranges::__require_access(__cgh, __rngs...);
+        auto __result_acc = __result_storage.__get_result_acc(__cgh);
+
+        __cgh.parallel_for<KernelName>(
+            sycl::nd_range</*dim=*/1>(sycl::range</*dim=*/1>(__wgroup_size), sycl::range</*dim=*/1>(__wgroup_size)),
+            [=](sycl::nd_item</*dim=*/1> __item_id) {
+                auto __local_idx = __item_id.get_local_id(0);
+
+                // 1. Set initial value to local found state
+                _AtomicType __found_local = __init_value;
+
+                // 2. Find any element that satisfies pred
+                //  - after this call __found_local may still have initial value:
+                //    1) if no element satisfies pred;
+                //    2) early exit from sub-group occurred: in this case the state of __found_local will updated in the next group operation (3)
+                __pred(__item_id, __rng_n, __iters_per_work_item, __wgroup_size, __found_local, __brick_tag, __rngs...);
+
+                // 3. Reduce over group: find __dpl_sycl::__minimum (for the __parallel_find_forward_tag),
+                // find __dpl_sycl::__maximum (for the __parallel_find_backward_tag)
+                // or update state with __dpl_sycl::__any_of_group (for the __parallel_or_tag)
+                // inside all our group items
+                if constexpr (__or_tag_check)
+                    __found_local = __dpl_sycl::__any_of_group(__item_id.get_group(), __found_local);
+                else
+                    __found_local = __dpl_sycl::__reduce_over_group(__item_id.get_group(), __found_local,
+                                                                    typename _BrickTag::_LocalResultsReduceOp{});
+
+                // Set local found state value value to global state to have correct result
+                if (__local_idx == 0)
+                {
+                    _AtomicType& __found =
+                        *__result_and_scratch_storage_t::__get_usm_or_buffer_accessor_ptr(__result_acc);
+
+                    __found = __found_local;
```

Comment:
I don't know if this works if `__result_and_scratch_storage_t` is in buffer storage mode.  This would return an accessor, which doesn't overload the `operator*()`, I believe.

I think the better way is to not dereference, but rather assign to an `auto` typed variable, then assign `__found_local` to the zeroth element of the accessor / usm ptr.

Reply:
As I may see below, the function `__get_usm_or_buffer_accessor_ptr` always return pointer:
```C++
    template <typename _Acc>
    static auto
    __get_usm_or_buffer_accessor_ptr(const _Acc& __acc, ::std::size_t __scratch_n = 0)
    {
#if _ONEDPL_SYCL_UNIFIED_USM_BUFFER_PRESENT
        return __acc.__get_pointer();
#else
        return &__acc[__scratch_n];
#endif
    }
```

So we simple apply `operator*` to this pointer and save result into reference:
```C++
                    __FoundStateType& __found =
                        *__result_and_scratch_storage_t::__get_usm_or_buffer_accessor_ptr(__result_acc);
```

Looks like this code hasn't problems here...

Reply:
I agree with Dan. Assigning the value to the zeroth element seems the safer choice here and is used in our existing code already. If you want to stick with this code, please test it with the buffer mode by setting `_ONEDPL_SYCL_UNIFIED_USM_BUFFER_PRESENT` to false or use an older compiler. Another path you would need to try is setting `__use_USM_host` and `__supports_USM_device` to false in `__result_and_scratch_storage`.

Reply:
`sycl::accessor` based on `sycl::buffer` return in this case `reference` type.
After that inside the method
```C++
    auto
    __get_pointer() const // should be cached within a kernel
    {
        return __usm ? __ptr + __offset : &__acc[__offset];
    }
```
we return address from this reference.
So I think it's look ok too.

Reply:
About the other our cases, for example `__work_group_reduce_kernel` - there is double referencing: we pass pointer into `__work_group_reduce_kernel` by const reference, names it as `__res_acc` and then, as you wrote, assign value to in through `[0]'. For me this practice isn't very good.

Reply:
Fixed: I changed this place for compatibility with other code, but not because it's incorrect.

Reply:
Ah, I guess I misunderstood how `__get_usm_or_buffer_accessor_ptr()` worked, but your logic is correct I think @SergeyKopienko . 

The naming of the function gives it away I suppose, a pointer to a buffer accessor.   Its just not a normal thing to do, but it does allow the return to be treated equivalently to the usm pointer, I guess.

I think its more clear as you have it written now, but I'm OK with either option.